### PR TITLE
JENKINS-68116 Export metrics for SCM Event queue (req Jenkins 2.289.3)

### PR DESCRIPTION
**Pull request by timja**
I would like to be able to monitor our SCM event queue and be able to measure the effect that tuning I do has on it.

see https://issues.jenkins.io/browse/JENKINS-68116

requires https://github.com/jenkinsci/scm-api-plugin/pull/127

image

### CODE REVIEW

**Review by kuisathaverat — APPROVED** [review]

**Review by v1v — APPROVED** [review]
💯 Thanks Tim

File: pom.xml [hunk 1]
```diff
@@ -263,6 +263,11 @@
             <groupId>org.jenkins-ci.plugins</groupId>
             <artifactId>pipeline-stage-step</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>scm-api</artifactId>
+            <version>999999-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (timja) [reply pom.xml:269]:
```suggestion
            <version>596.v47ec118b_0c3a_</version>
```

**Comment by cyrille-leclerc** [comment]
Bumping from 2.277 to 2.289 would block the upgrade of 8 installations out of 173 reported installs. 5% of the install base. that may be ok as our current version looks pretty stable.

**Comment by cyrille-leclerc** [comment]
@timja is it an activity that should also be captured as part of the traces of the job executions? Are there APIs to hook before and after this SCM task to introduce a span in the traces?

**Comment by timja** [comment]
> Bumping from 2.277 to 2.289 would block the upgrade of 8 installations out of 173 reported installs. 5% of the install base. that may be ok as our current version looks pretty stable.

required by scm-api :)

**Comment by cyrille-leclerc** [comment]
> > Bumping from 2.277 to 2.289 would block the upgrade of 8 installations out of 173 reported installs. 5% of the install base. that may be ok as our current version looks pretty stable.
> 
> required by scm-api :)

I don't deny :-) it's always complicated to find the right trade off. Better monitoring is definitively a compelling reason. We would have to ensure we maximise this ROI providing the best guided workflows to get benefits of this bump.

> > I imagine it would be great to see these SCM activities as spans with all the contextual attributes (repository url, user...) to slice and dice the slowness in any dimension and understand if it's specific to some scm repositories or if it impact all repositories.
> 
> I'm not sure, an event can map to many jobs starting

Ack thanks

**Comment by timja** [comment]
Converting to draft till upstream is merged

File: src/main/java/io/jenkins/plugins/opentelemetry/init/SCMEventMonitoringInitializer.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.logging.Logger;
+import javax.annotation.Nonnull;
+import javax.inject.Inject;
+import jenkins.YesNoMaybe;
+import jenkins.scm.api.SCMEvent;
+
+/**
+ * Capture SCM Events metrics
+ */
+@Extension(dynamicLoadable = YesNoMaybe.YES, optional = true)
```
Comment (jtnord) [inline src/main/java/io/jenkins/plugins/opentelemetry/init/SCMEventMonitoringInitializer.java:24]: > (dynamicLoadable = YesNoMaybe.YES)

does the @Initializer get called if this is installed when Jenkins is already running?

Comment (timja) [reply src/main/java/io/jenkins/plugins/opentelemetry/init/SCMEventMonitoringInitializer.java:24]: I don't think so, same issue applies to a few other metrics, you need to restart it unfortunately.

Comment (timja) [reply src/main/java/io/jenkins/plugins/opentelemetry/init/SCMEventMonitoringInitializer.java:24]: Possibly, would rather just fix that though.

You don't happen to know how to init on plugin install dynamically loading only?

Comment (jtnord) [reply src/main/java/io/jenkins/plugins/opentelemetry/init/SCMEventMonitoringInitializer.java:24]: rather hacky `Plugin.start()` is called so you can have a `start` method that checks the current init state and if Jenkins init state is complete then you know the plugin is being dynamically loaded and you can take some action.
(you can probably do the same with an `Extension` in the constructor (if the constructor is called when Jenkins is "fully up" then it has been installed post startup).

Seems rather hacky.  Would possibly be better to fix `PluginManager.dynamicLoad()` (if indeed Initializers are not called from some other place).

Comment (cyrille-leclerc) [reply src/main/java/io/jenkins/plugins/opentelemetry/init/SCMEventMonitoringInitializer.java:24]: Interesting. I thought that marking an Extension as dynamically loadable made it loaded during the install as well as at startup time.

Comment (jtnord) [reply src/main/java/io/jenkins/plugins/opentelemetry/init/SCMEventMonitoringInitializer.java:24]: The extension will be loaded but the Initializer may not be called, so you can end up with the extension running but in an incorrect state for what it expects.

Comment (timja) [reply src/main/java/io/jenkins/plugins/opentelemetry/init/SCMEventMonitoringInitializer.java:24]: Seems this works just fine, installing a plugin re-runs the init milestones, I verified in the debugger it gets called:

```
2022-04-19 10:40:22.223+0000 [id=114]	INFO	h.m.UpdateCenter$CompleteBatchJob#run: Completing installing of plugin batch…
2022-04-19 10:40:22.777+0000 [id=121]	INFO	jenkins.InitReactorRunner$1#onAttained: Started initialization
2022-04-19 10:40:22.780+0000 [id=126]	INFO	jenkins.InitReactorRunner$1#onAttained: Listed all plugins
2022-04-19 10:40:22.781+0000 [id=131]	INFO	jenkins.InitReactorRunner$1#onAttained: Prepared all plugins
2022-04-19 10:40:22.784+0000 [id=135]	INFO	jenkins.InitReactorRunner$1#onAttained: Started all plugins
2022-04-19 10:40:22.784+0000 [id=121]	INFO	jenkins.InitReactorRunner$1#onAttained: Augmented all extensions
2022-04-19 10:40:22.784+0000 [id=126]	INFO	jenkins.InitReactorRunner$1#onAttained: System config loaded
2022-04-19 10:40:22.784+0000 [id=131]	INFO	jenkins.InitReactorRunner$1#onAttained: System config adapted
2022-04-19 10:40:22.785+0000 [id=123]	INFO	jenkins.InitReactorRunner$1#onAttained: Loaded all jobs
2022-04-19 10:40:22.785+0000 [id=128]	INFO	jenkins.InitReactorRunner$1#onAttained: Configuration for all jobs updated
2022-04-19 10:40:22.785+0000 [id=132]	INFO	jenkins.InitReactorRunner$1#onAttained: Completed initialization
2022-04-19 10:40:22.790+0000 [id=114]	INFO	h.m.UpdateCenter$CompleteBatchJob#run: Completed installation of 1 plugins in 3.2 sec
```

**Comment by timja** [comment]
cc @jenkinsci/opentelemetry-plugin-developers

**Review by kuisathaverat — APPROVED** [review]
LGTM

**Pull request merged**